front/components/assistant_builder/InstructionScreen.tsx

```diff
@@ -126,19 +130,63 @@ export function InstructionScreen({
         limit: INSTRUCTIONS_MAXIMUM_CHARACTER_COUNT,
       }),
     ],
-    content: tipTapContentFromPlainText(builderState.instructions || ""),
+    editable: !doTypewriterEffect,
+    content: tipTapContentFromPlainText(
+      (!doTypewriterEffect && builderState.instructions) || ""
+    ),
     onUpdate: ({ editor }) => {
-      const json = editor.getJSON();
-      const plainText = plainTextFromTipTapContent(json);
-      setEdited(true);
-      setBuilderState((state) => ({
-        ...state,
-        instructions: plainText,
-      }));
+      if (!doTypewriterEffect) {
+        const json = editor.getJSON();
+        const plainText = plainTextFromTipTapContent(json);
+        setEdited(true);
+        setBuilderState((state) => ({
+          ...state,
+          instructions: plainText,
+        }));
+      }
     },
   });
   const editorService = useInstructionEditorService(editor);
 
+  const [letterIndex, setLetterIndex] = useState(0);
+
+  useEffect(() => {
```

Comment:
Looking more at this, it's pretty heavy re-rendering. For each letter we will re-render this component. Either you move this outside of this component so the useEffect is not part of a rendered component but only uses the TipTap API to play the animation. This use case is also often achieve using `requestanimationframe`. Also I'm wondering if we could not leverage CSS styles here, with something like this:

```
    editor.chain()
      .setContent(`<span class="typewriter-effect">${textContent}</span>`)
      .focus("end")
      .run();

    // Set editable to false while typewriter effect is running
    editor.setEditable(false);

    // Once the animation is complete, set editable to true
    const animationDuration = 4000; // Duration should match the CSS animation duration
    const timeoutId = setTimeout(() => {
      editor.setEditable(true);
      setDoTypewriterEffect(false);
    }, animationDuration);
```

Reply:
Added some comments and we transcript our findings in PR description.